python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -0,0 +1,227 @@
+# Copyright (c) 2020-2023, NVIDIA CORPORATION.
+from typing import Any, Dict
+
+import numba
+import pandas as pd
+from numba import cuda, types
+from numba.core.extending import (
+    make_attribute_wrapper,
+    models,
+    register_model,
+    type_callable,
+    typeof_impl,
+)
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AbstractTemplate, AttributeTemplate
+from numba.cuda.cudadecl import registry as cuda_registry
+from numba.np import numpy_support
+
+index_default_type = numpy_support.from_dtype(
+    pd.RangeIndex(0, 0).dtype
+)  # int64
+SUPPORTED_GROUPBY_NUMBA_TYPES = [types.int64, types.float64]
+SUPPORTED_GROUPBY_NUMPY_TYPES = [
+    numpy_support.as_dtype(dt) for dt in [types.int64, types.float64]
+]
+
+
+class Group(object):
+    """
+    A piece of python code whose purpose is to be replaced
+    during compilation. After being registered to GroupType,
+    serves as a handle for instantiating GroupType objects
+    in python code and accessing their attributes
+    """
+
+    def __init__(self, group_data, size, index, dtype, index_dtype):
+        self.group_data = group_data
+        self.size = size
+        self.index = index
+        self.dtype = dtype
+        self.index_dtype = index_dtype
```

Comment:
Two questions:
1. Are the types of these things all known, or could some of them vary (e.g. is dtype always a dtype object, or could it still be some arbitrary string like "float")?
2. Does numba support [dataclasses](https://docs.python.org/3/library/dataclasses.html)?

Depending on the answer Group would be a good candidate for a dataclass.

Reply:
I seems we don't need this attributes at all anymore. This class only exists so that we can have something to overload in our template kernel that will get replaced with a `GroupType`. Maybe at some point we needed to write `Group.size` in our template kernel outside of the UDF's invocation, but in any case the tests seem to pass without these at all so I just removed them.